.github/workflows/builder_docker-based_slsa3.yml

```diff
@@ -208,13 +220,13 @@ jobs:
             --build-config-path "${CONFIG_PATH}" \
             --build-definition-path build-definition.json \
             --builder-image "${BUILDER_IMAGE}@${BUILDER_DIGEST}" \
-            --git-commit-digest "sha1:${GITHUB_SHA}" \
+            --git-commit-digest "sha1:${{ github.event.pull_request.head.sha }}" \
```

Comment:
With this change we are getting the correct commit sha, but I think we also need to specify the correct branch in `source-repo`. I am not sure if that is possible. I'll give it a try.

Reply:
IIUC the head SHA will only be available for `pull_request` events? Are there other events you'll need to handle? I assume that the primary event you would need to handle is `tag` or `release` for when release tags are created.

Reply:
Yes. We have to be able to handle all these events. I am thinking of reverting this back to use `GITHUB_SHA`, but make these steps conditional on events that are not `pull_request`, and add a separate step for `pull_request`.

Reply:
Yeah, we have a lot of places throughout that act differently for pull requests.

I'm curious how you are planning to support pull requests? we have the problem that `id-token:write` isn't available on pull requests (https://github.com/slsa-framework/slsa-github-generator/issues/131) so we aren't really able to sign provenance for those events currently.

Reply:
For pull-requests we only run the `dry-run` command. We won't actually generate the provenance or sign, but just to make sure that the inputs are correct, and that we can create a BuildDefinition. 

The idea is to allow the users to verify their workflow parameters before merging.

Reply:
> but make these steps conditional on events that are not pull_request, and add a separate step for pull_request.

+1

@ianlewis also so far we can also do this:
https://github.com/slsa-framework/slsa-github-generator/blob/76f03fa7e30209f32ac76ce417ddc43ff98af42a/.github/workflows/builder_docker-based_slsa3.yml#L396

since the signing is isolated in a step. in pull_request we just published the unsigned attestations.

there's a question of whether there are other events besides pull_request that don't have id-token. Perhaps we should actually replace the `if pull_requests` to `if -n $ACTIONS_ID_TOKEN_REQUEST_URL`.

Reply:
You'll at least need to detect pull request actions so you know what sha to use so it might be best to just use that.

If the code for each is the same outside of deciding what sha to use you could set the correct sha in a prior build step.

```yaml
- name: Detect SHA
  id: get-sha
  env:
    GITHUB_HEAD_SHA: "${{ github.event.pull_request.head.sha }}"
  run: |
    if [ "$GITHUB_EVENT_NAME" == "pull_request" ]; then
      echo "sha=$GITHUB_HEAD_SHA" >> $GITHUB_OUTPUT
    else
      echo "sha=$GITHUB_SHA" >> $GITHUB_OUTPUT
    fi
```

and then use it

`${{ steps.get-sha.outputs.sha }}`

Reply:
I am not sure if it is even possible to make it work on pull_requests. The issue is not just about the commit sha, but also the repo-url. I am not sure what the repo-url could be in the case of a pull-request.

So I am thinking of a completely different approach:

The first thing that happens during [the setup step](https://github.com/slsa-framework/slsa-github-generator/blob/76f03fa7e30209f32ac76ce417ddc43ff98af42a/internal/builders/docker/pkg/builder.go#L139) is to check if the latest commit matches the given commit (this happens in [`verifyCommit`](https://github.com/slsa-framework/slsa-github-generator/blob/76f03fa7e30209f32ac76ce417ddc43ff98af42a/internal/builders/docker/pkg/builder.go#L303)). If the commits don't match, an error is returned, unless `force-checkout` is provided in which case the repo is cloned. 

I am thinking of allowing a third option. When `verifyCommit` detects a commit that does not match the input commit, we could do either of:
1. Abort. This should be the default option.
2. Force checkout.
3. Use the existing checked out commit as the source. This option only makes sense if the repo is checked out from the same repo that is given in the repo-url. We can check this using `git remote get-url origin` (this command only works if there is a remote named `origin`). 

So I am thinking of replacing the `--force-checkout` option with another flag, e.g., `--unexpected-commit` which could take one of these three values: `abort`, `force-checkout` and `ignore`; with `abort` being the default. 

This way, for `dry-run` we could use `--unexpected-commit=ignore` to allow generating the `BuildDefinition` using the config file on the current working directory even in the case of a `pull-request` without having to clone the repo.

What do you think?

I could give it a try in this PR.

Reply:
I think that makes sense as long as it's only an option available on the dry run

Reply:
Perfect. I'll give it a try.

Reply:
> and the problem with pull-request events was that the repository was not checked out for the dry-run workflow. 

OH NO!

OK so just to clarify what happens here now: `actions/checkout` by default checks out the github repo at the ref of the merge commit. I think that also references `GITHUB_SHA`, and so the comparison with the checked out repo and the `GITHUB_SHA` of the command line input should always match, pull request or not?

secondly, the new `resolution-strategy` will allow dry-run to create a build definition, ignoring the current checkout and using the command line provided git-commit-digest? if so, i think that makes sense.

what happens to the digest of the config file though? is it checked out at the current checkout or at the git-commit-digest during `--resolution-strategy ignore`? (It seems like `ignore` is like `override` with the current checkout so it uses the current checkout for the config file)

Reply:
> OK so just to clarify what happens here now: `actions/checkout` by default checks out the github repo at the ref of the merge commit. I think that also references `GITHUB_SHA`, and so the comparison with the checked out repo and the `GITHUB_SHA` of the command line input should always match, pull request or not?

Yes. After all these experiments that is my understanding too. This behaviour makes sense, and makes things simpler :)

> secondly, the new `resolution-strategy` will allow dry-run to create a build definition, ignoring the mismatch and using the current checkout for the source digest? if so, i think that makes sense.

True. But actually this feature is no longer required, at least not for GitHub. So we could either keep it, or drop it in favour of making the code simpler :)

> what happens to the digest of the config file though? is it checked out at the current checkout or at the git-commit-digest during `--resolution-strategy ignore`? (It seems like `ignore` is like `override` with the current checkout so it uses the current checkout for the config file)

It is checked out at the current commit, or what I call the `effective commit`. The generated `BuildDefinition` also contains this effective commit digest, not the one given in the inputs.

Reply:
> So we could either keep it, or drop it in favour of making the code simpler :)

Should we keep, and file a follow-up issue to clean it if the code is complete and we find it's not useful for testing?

Reply:
I can actually drop these changes easily, I just need to revert the last commit.

Reply:
easy! thank you!

Reply:
Done. I think this is now ready to be merged :)